### PR TITLE
*: Log SQL statement when coprocessor encounteres lock

### LINKED ISSUE #27718: Show information about SQL statement in the debug log about coprocessor encountering lock
## Enhancement

When coprocessor reading meets lock, TiDB will print a [debug log](https://github.com/pingcap/tidb/blob/477593e31f281d57d66319a66d56326baf809c1a/store/copr/coprocessor.go#L885). However it shows nothing about the related SQL statements, and there's not anywhere else that shows the statements that's affected by the lock. If this log can print the statement, it will be helpful for diagnose some problems. Of course, the statement in the log should be redacted according to the security options.

### PR DESCRIPTION
**Pull request by MyonKeminta**
### What problem does this PR solve?

Issue Number: close #27718

Problem Summary: The debug log on coprocessor encounters lock doesn't print out the statement, which makes it difficult to know what statement was affected by the lock. This PR tries to add the log.

### What is changed and how it works?

What's Changed: Added a callback parameter to the `Send` method of `kv.Client` that accepts a new TransactionEvent object. This avoids to pass the SQL statement deep down to the coprocessor client part. And I think this callback and events can also be reused by Lock View.

How it Works:

### Check List 

Tests 

- [ ] Manual test (add detailed scripts or steps below) (TODO)

Side effects

- [ ] Performance regression: Consumes more CPU

Documentation

- None

### Release note 

```release-note
None.
```

### CODE REVIEW

File: util/trxevents/trx_events.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+// CopMeetLock represents an event that coprocessor reading encounters lock.
+type CopMeetLock struct {
+	LockInfo *kvrpcpb.LockInfo
+}
+
+// TransactionEvent represents a transaction event that may belong to any of the possible types.
+type TransactionEvent struct {
+	eventType EventType
+	ptr       unsafe.Pointer
```
Comment (longfangsong) [inline util/trxevents/trx_events.go:39]: Why we have to use `unsafe.Pointer` here?

**Review by longfangsong — APPROVED** [review]
Just a small question. Rest LGTM

**Comment by longfangsong** [comment]
> @longfangsong: Thanks for your review. The bot only counts approvals from reviewers and higher roles in [list](https://prow.tidb.io/tichi/repos/pingcap/tidb/pulls/27735/owners), but you're still welcome to leave your comments.

PTAL [pingcap/community#544](https://github.com/pingcap/community/pull/544)

File: util/trxevents/trx_events.go (hunk 1 above)
Comment (MyonKeminta) [reply util/trxevents/trx_events.go:39]: It seems either can be used, but interface{} looks better.  I thought about avoiding the RTTI (though I'm not sure if there is) but I think perhaps that's a bad idea. I'll change it to interface{} then, please wait a minute

**Comment by MyonKeminta** [comment]
@sticnarf @cfzjywxk Also PTAL, thanks

**Review by cfzjywxk — APPROVED** [review]

**Review by sticnarf — APPROVED** [review]
LGTM

**Pull request merged**